kubeadm: update the 1.14 HA guide (#13191)
kubernetes/website: content/en/docs/setup/independent/high-availability.md

Before proceeding, you should carefully consider which approach best meets the needs of your applications
and environment. [This comparison topic](/docs/setup/independent/ha-topology/) outlines the advantages and disadvantages of each.

You should also be aware that setting up HA clusters with kubeadm is still experimental and will be further
simplified in future versions. You might encounter issues with upgrading your clusters, for example.
We encourage you to try either approach, and provide us with feedback in the kubeadm
[issue tracker](https://github.com/kubernetes/kubeadm/issues/new).

> See also [The HA upgrade documentation](/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade-ha-1-14).

Reviewer: Isn't the HA upgrade documentation being consolidated with the standard upgrade documentation?

        apiVersion: kubeadm.k8s.io/v1beta1
        kind: ClusterConfiguration

>         kubernetesVersion: stable

Reviewer: Is this SAN generated automatically now? I know in the past it wasn't.

Reviewer 2: yes, the code is [here](https://github.com/kubernetes/kubernetes/blob/df2094b3d728bd58c5f23b41add109e32fc7c301/cmd/kubeadm/app/util/pkiutil/pki_helpers.go#L386)

        controlPlaneEndpoint: "LOAD_BALANCER_DNS:LOAD_BALANCER_PORT"

    - `kubernetesVersion` should be set to the Kubernetes version to use. This
  example uses `stable`.
    - `controlPlaneEndpoint` should match the address or DNS and port of the load balancer.
    - It's recommended that the versions of kubeadm, kubelet, kubectl and Kubernetes match.

    {{< note >}}

>     Some CNI network plugins like Calico require a CIDR such as `192.168.0.0/16` and

Reviewer 2: In preview this is rendering as code 🤔

Reviewer 2: Might be notes should not be indented?

      kubeadm join 192.168.0.200:6443 --token 9vr73a.a8uxyaju799qwdjv --discovery-token-ca-cert-hash sha256:7c2e69131a36ae2a042a339b33381c6d0d43887e2de83720eff5359e26aec866 --experimental-control-plane --certificate-key f8902e114ef118304e561c3ecd4d0b543adc226b7a07f675f56564185ffe0c07
    
    Please note that the certificate-key gives access to cluster sensitive data, keep it secret!
    As a safeguard, uploaded-certs will be deleted in two hours; If necessary, you can use kubeadm init phase upload-certs to reload certs afterward.

    Then you can join any number of worker nodes by running the following on each as root:
      kubeadm join 192.168.0.200:6443 --token 9vr73a.a8uxyaju799qwdjv --discovery-token-ca-cert-hash sha256:7c2e69131a36ae2a042a339b33381c6d0d43887e2de83720eff5359e26aec866
    ```

    - Copy this output to a text file. You will need it later to join control plane and worker nodes to the cluster.
    - When `--experimental-upload-certs` is used with `kubeadm init`, the certificates of the primary control plane
    are encrypted and uploaded in the `kubeadm-certs` Secret.

>     - Please note that the `kubeadm-certs` Secret and decryption key expire after two hours. To re-upload the certificates

Reviewer: This should probably be called out in a note tag to make it more obvious.

    and generate a new decryption key, use the following command on a control plane that is already joined the cluster:

      ```sh
      sudo kubeadm init phase upload-certs --experimental-upload-certs
      ```

>     {{< caution >}}

Reviewer 2: Same here

>     and generate a new decryption key, use the following command on a control plane that is already joined the cluster:

Reviewer: ```suggestion
    and generate a new decryption key, use the following command on a control plane node that is already joined to the cluster:
```

As stated in the command output, please note that the certificate-key gives access to cluster sensitive data, keep it secret!
{{< /caution >}}

1.  Apply the CNI plugin of your choice:

    [Follow these instructions](/docs/setup/independent/create-cluster-kubeadm/#pod-network) to install

>     the pod network. Make sure this corresponds to whichever pod CIDR you provided in the kubeadm configuration file.

Reviewer: ```suggestion
    the CNI provider. Make sure the configuration corresponds to the pod CIDR specified in the kubeadm configuration file if applicable.
```

                certFile: /etc/kubernetes/pki/apiserver-etcd-client.crt
                keyFile: /etc/kubernetes/pki/apiserver-etcd-client.key

>       {{< note >}}

Reviewer 2: Same here

The steps are the same as for the stacked etcd setup:

- Make sure the first control plane node is fully initialized.
- Join each control plane node with the join command you saved to a text file. It's recommended
to join the control plane nodes one at a time.
- Don't forget that the decryption key from `--certificate-key` expires after two hours, by default.

## Common tasks after bootstrapping control plane

### Install workers

Worker nodes can be joined to the cluster with the command you stored previously
as the output from the `kubeadm init` command:

```sh
sudo kubeadm join 192.168.0.200:6443 --token 9vr73a.a8uxyaju799qwdjv --discovery-token-ca-cert-hash sha256:7c2e69131a36ae2a042a339b33381c6d0d43887e2de83720eff5359e26aec866
```

## Manual certificate distribution {#manual-certs}

If you choose to not use `kubeadm init` with the `--experimental-upload-certs` flag this means that
you are going to have to manually copy the certificates from the primary control plane node to the
joining control plane nodes.

There are many ways to do this. In the following example we are using `ssh` and `scp`:

SSH is required if you want to control all nodes from a single machine.

> 1.  Enable ssh-agent on your main device that has access to all other nodes in

Reviewer 3: Would it be simpler to document using `scp -3` and skip using an SSH agent?

(`scp -3` runs all transfers via the local client).

Author: in the case of `scp -3` routing we do avoid the need for ssh-agent.
but the idea of enabling ssh-agent is to also allow to ssh to the nodes and execute commands on them.
